core/templates/dev/head/editor/EditorServices.js

```diff
@@ -1836,22 +1846,23 @@ oppia.factory('explorationWarningsService', [
           _graphData.finalStateIds, _graphData.nodes,
           _getReversedLinks(_graphData.links));
         if (deadEndStates.length) {
-          var deadEndStatesString = null;
-          if (deadEndStates.length === 1) {
-            deadEndStatesString = 'the following card: ' + deadEndStates[0];
-          } else {
-            deadEndStatesString = 'each of: ' + deadEndStates.join(', ');
-          }
-
-          _warningsList.push({
-            type: WARNING_TYPES.ERROR,
-            message: (
-              'Please make sure there\'s a way to complete ' +
-              'the exploration starting from ' +
-              deadEndStatesString + '.')
+          angular.forEach(deadEndStates, function(deadEndState) {
+            if (_statesWithErrors.indexOf(deadEndState) === -1) {
+              _statesWithErrors.push(deadEndState);
+            }
+            _perStateErrors.push({
+              type: WARNING_TYPES.ERROR,
+              message: ('Please make sure there\'s a way to complete ' +
+              'the exploration starting from this card: ' + deadEndState)});
           });
         }
       }
+      _warningsList.push({
+            type: WARNING_TYPES.ERROR,
+            message: (
+              'The following states have errors: ' +
```

Comment:
With the new data structure for _perStateErrors, I don't think we need to separately keep track of _statesWithErrors anymore. We can just do Object.keys(_perStateErrors) to get the list of state names.